### PR TITLE
Stream response bodies instead of always buffering them completely

### PR DESCRIPTION
**Pull request by kelunik**
| Q               | A
| --------------- | ---
| Bug fix?        | kinda
| New feature?    | kinda
| BC breaks?      | unknown
| Deprecations?   | no
| Related tickets | Mentioned in #3.
| License         | MIT

#### What's in this PR?

Use a custom `StreamInterface` implementation to allow for proper backpressure.

#### Why?

Large responses have to be held entirely in memory otherwise.

### CODE REVIEW

**Review by Nyholm — APPROVED** [review]

**Comment by Nyholm** [comment]
Why do you need an anonymous class btw?
It will be easier to test if it isn't.

**Comment by kelunik** [comment]
Doesn't need to be one, but as there's only a single use I just did that.

File: src/Client.php [hunk 1]
```diff
@@ -40,19 +46,140 @@ public function sendRequest(RequestInterface $request)
...
+                        }
+
+                        if ($this->buffer === null) {
+                            $this->eof = true;
+
+                            return '';
+                        }
+                    }
+
+                    $read = \substr($this->buffer, 0, $length);
```
Comment (joelwurtz) [inline src/Client.php]: what happens when the length is higher that the buffer size, it returns less ? not sure if this is compliant with psr7.

File: src/Client.php [hunk 2]
```diff
@@ -40,19 +46,140 @@ public function sendRequest(RequestInterface $request)
...
+                    $this->buffer = (string) \substr($this->buffer, $length);
+                    $this->position += \strlen($read);
```
Comment (joelwurtz) [inline src/Client.php]: not sure if we need the position as we cannot seek this stream it's also pretty useless IMO (but doesn't hurt, so don't really care)

File: src/Client.php [hunk 3]
```diff
@@ -40,19 +46,140 @@ public function sendRequest(RequestInterface $request)
...
+                }
+
+                public function __destruct()
+                {
+                    $this->close();
+                }
+
+                public function close()
+                {
+                    $this->cancellationTokenSource->cancel();
```
Comment (joelwurtz) [inline src/Client.php]: not a pro on amp, it does not give problems if the stream is already closed to cancel it twice ?

Comment (bwoebi) [reply src/Client.php]: Nope, it's guarded and will just immediately return when called multiply. (also: if called after regular close, the handler attached to the cancellationTokenSource will have been detached and have no effect either)

**Review by dbu — COMMENTED** [review]
if i read this correctly, we still need to copy the request but no longer need to copy the response. that looks very good!

i would prefer to move the stream implementation to its own class for better readability. and then you could also add a few unit tests for the stream to ensure it works as expected.

**Comment by fbourigault** [comment]
I'm also not a big fan of the anonymous class. It make things harder to read and to test.

File: src/Client.php (hunk 3 above)
Comment (kelunik) [reply src/Client.php]: Promises have that protection, but it doesn't make sense to have it for cancellation, so no as @bwoebi already explained.

File: src/Client.php (hunk 1 above)
Comment (kelunik) [reply src/Client.php]: It returns less, right. PSR-7 defines:

> Read up to $length bytes from the object and return them. Fewer than $length bytes may be returned if underlying stream call returns fewer bytes.

File: src/Client.php (hunk 2 above)
Comment (kelunik) [reply src/Client.php]: True, but `tell` is also used for determining how many bytes have been read without seeking.

**Comment by kelunik** [comment]
@dbu We can implement streaming for the request pretty easily as long as we keep the "never follow redirects", otherwise we can only stream for rewindable streams, because redirects might require sending the body again.

@dbu @fbourigault What's the name you suggest for that class? Which areas specifically _need_ more tests? IMO everything that's tested should be tested by the common tests, not with individual tests in different clients.

**Comment by dbu** [comment]
> We can implement streaming for the request pretty easily as long as we keep the "never follow redirects", otherwise we can only stream for rewindable streams, because redirects might require sending the body again.

i suggest adding an issue that there is a possible performance improvement of checking if the stream is rewindable and only then copy the request. but not do it now, so we can get a first release ready. its only relevant when sending large requests.

> What's the name you suggest for that class? Which areas specifically need more tests? IMO everything that's tested should be tested by the common tests, not with individual tests in different clients.

ArtaxStreamAdapter or something like that? the common tests do not test edge cases of streams i suppose, adding such tests would be good to know exactly when something in the stream is broken, as opposed to "something" with the client somewhere... e.g. when we implement the first point, we can have different tests for rewindable and non rewindable streams and ensure the behaviour is always correct.

**Comment by Nyholm** [comment]
FYI: There are some unit tests for PSR7 streams in https://github.com/php-http/psr7-integration-tests

**Comment by kelunik** [comment]
@Nyholm Unfortunately, the stream tests there assume `StreamInterface` to be tightly coupled to PHP's resources.

What should `getContents()` return on a second call? According to the PSR document, it returns the remaining contents, but it doesn't say whether they're consumed or just returned. So should the next invocation return the same thing or just an empty string?

**Comment by Nyholm** [comment]
Hm, maybe they are. We should review that. 


>What should getContents() return on a second call? According to the PSR document, it returns the remaining contents, but it doesn't say whether they're consumed or just returned. So should the next invocation return the same thing or just an empty string? 

It should return an empty string.

**Comment by kelunik** [comment]
@Nyholm Moved it to a separate class and added tests for full coverage.

File: src/Internal/ResponseStream.php [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+use Amp\Artax;
+use Amp\ByteStream\InputStream;
+use Amp\ByteStream\IteratorStream;
+use Amp\CancellationTokenSource;
+use Amp\CancelledException;
+use Amp\Emitter;
+use Amp\Promise;
+use Psr\Http\Message\StreamInterface;
+
+/** @internal */
```
Comment (dbu) [inline src/Internal/ResponseStream.php]: i would love a short comment what this is. something like "Stream wrapper to expose the amp stream as PSR-7 compatible stream."

**Review by dbu — APPROVED** [review]
this looks good to me, thanks!

File: src/Internal/ResponseStream.php (hunk 4 above)
Comment (kelunik) [reply src/Internal/ResponseStream.php]: Done.

**Comment by dbu** [comment]
cheers!

**Comment by Nyholm** [comment]
Great work!
